### ISSUE TITLE
Enhancement: Disallow assigning while declaring

### ISSUE THREAD

**Issue by mietek** [koalaman/shellcheck#262]
Freshly debugged:

``` diff
@@ -570,7 +570,8 @@ restore_updated_cabal_layer () {

    log 'Locating Cabal layers'

-   local updated_name=$(
+   local updated_name
+   updated_name=$(
        list_stored_files "${platform}/${archive_prefix}" |
        sed "s:^${platform}/::" |
        match_updated_cabal_archive_name "${tag}"
```

It would be great if ShellCheck could flag any case where a variable has a value assigned at the same time it is declared.  Doing this loses the exit status of any command used to generate the value, as `$?` will now refer to the successful exit status of `local`.

**Comment by ormaaj** [comment]
It's true that's a pitfall, though it would really need to be special cased to only warn when something depends on the status. Declaring and assigning together is way too common practice to apply it generally.  It would generate false alarms more often than not.

**Comment by chelmertz** [comment]
A while ago I spent a bit too much time to realize this, after having used shellcheck on the script. I learned to avoid it and look for it elsewhere, but I think shellcheck should have caught it.

>It's true that's a pitfall [...]

That's why I use shellcheck, to avoid pitfalls. I feel the whole point of this tool is to catch these corner cases. Ignoring some arbitrarily selected anti patterns because it would cause (old?) scripts to fail this tool's check does not say that the script does not work as expected in your environment, but I don't see why this tool should adapt.

>Declaring and assigning together is way too common practice to apply it generally [..]

We're only talking about using `local` here, which apparently causes bugs. For comparison, the tool reports on the use of expr which is perfectly valid in some cases but "antiquated": https://github.com/koalaman/shellcheck/wiki/SC2003

**Comment by koalaman** [comment]
@chelmertz so you had bug due to this too? Can you describe the circumstances?

Shellcheck should definitely be catching this, the question is more if it should be flagging all instances of `local `where a variable contains command substitution, or only when that `local ` appears as part of a conditional or in a script with `set -e`. 

@mietek, what do you think?

**Comment by chelmertz** [comment]
Yeah, I wanted the exit code after a command executed in a subshell. `local` + an assignment to something dynamic always turn this into 0 which I consider to be an anti pattern. That might not be a heavy argument though, since I am not experienced enough to have fallen into that trap earlier :)

I don't think `set -e` should represent a config input to shellcheck but if it already is I'm sure there's a reason for it.

**Comment by mietek** [comment]
I’d like to see warnings for all instances of `local` when the right-hand side includes command substitution. If someone is writing a quick’n’dirty script, they won’t be using `local` anyway.

It probably makes sense to do the same thing for `declare` and `export`, right?

**Comment by koalaman** [comment]
@chelmertz  @mietek 0e1a64b will warn about declare/export/local with command substitutions on RHS.

How do you feel about e.g. `foo=$(bar)$(baz)` where the return value of `bar` is masked by `baz`?

**Comment by mietek** [comment]
Thanks!  To answer your question, I think multiple command substitutions on the right-hand side would also merit a warning.

**Comment by mietek** [comment]
Perhaps this could be improved further, by recognising when the user means to ignore any return values:
```
In src/main.sh line 519:
		export BASHMENOT_TIMESTAMP_EPOCH=$( get_current_time || true )
                       ^-- SC2155: Declare and assign separately to avoid masking return values.
```

**Comment by refola** [comment]
Can you make ShellCheck only report "SC2155: Declare and assign separately to avoid masking return values." errors when it's used in a context where the return value is actually used? ShellCheck has really improved my Bash, but this "error" is happening a lot in my code which doesn't use return values. For example, [my btrfs snapshot and backup script](https://github.com/refola/scripts/blob/master/backup/backup-btrfs.sh) has the "error" 9 times, none of which use return values.

**Comment by bittorf** [comment]
me too. only complain if $? is really used.
otherwise it makes some functions harder to read,
because you must everytime write e.g.

```
local x
x="$( bla )"
```

**Comment by yoshuawuyts** [comment]
Yeah, :-1: on this as a whole. If you're doing iffy things with status codes, that's on you. For everyone that's trying to write imperative code like they would in any other scripting languages this makes it a lot harder to work with.

e.g. There's no reason this should ever give a warning:
```sh
do_something () {
  local date=$(echo "$@" | awk '{print $1}')
  echo "$date" | grep 2015
}
```

**Comment by chelmertz** [comment]
>If you're doing iffy things with status codes, that's on you

@yoshuawuyts That's one way to reason about a program that's there to *validate your mistakes*. You ever thought about those who unintentionally write bugs?

I'm outta here.

**Comment by bittorf** [comment]
@chelmertz read again: it's ok to warn, if the developer *wants* to do something with the returncode, otherwise the developer dont care if 'local' or the thing inside $(...) give the RC

**Comment by yoshuawuyts** [comment]
Yup, what @bittorf says. This rule is currently being enforced btw, even if returned status codes from local are not used

**Comment by mietek** [comment]
My suggestion is for ShellCheck to recognise `local=$( foo || true )` as an explicit admission that the status code is to be ignore.

**Comment by chelmertz** [comment]
@bittorf @yoshuawuyts Agreed; and I didn't pay much attention to the rest of your message, sorry. (And I guess the notifications bypass the unsubscribe-thingy :P)

**Comment by niieani** [comment]
I agree with @mietek. Aside from `|| true`, things like `var=$( foo || echo "whatever" )` should make ShellCheck ignore the output status code, since echo always returns 0.

**Comment by e-kwsm** [comment]
Now (0.11.0) shellcheck issues [SC2155](https://www.shellcheck.net/wiki/SC2155).
